music/src/core/sequences.ts

```diff
@@ -424,3 +424,129 @@ export function mergeInstruments(ns: INoteSequence) {
+ * `NoteSequences` in `args`. If not provided, the `totalTime` /
+ * `totalQuantizedSteps` of each sequence will be used. Specifying durations is
+ * useful if the sequences to be concatenated are effectively longer than
+ * their `totalTime` (e.g., a sequence that ends with a rest).
+ *
+ * @returns A new sequence that is the result of concatenating the input
+ * sequences.
+ */
+export function concatenate(
+    seqs: INoteSequence[], individualDuration?: number[]): NoteSequence {
+  if (individualDuration && individualDuration.length !== seqs.length) {
+    throw new Error(
+        // tslint:disable-next-line:max-line-length
+        'Number of sequences to concatenate and their individual durations does not match.');
+  }
+
+  if (isQuantizedSequence(seqs[0])) {
+    // Check that all the sequences are quantized, and that their quantization
+    // info matches.
+    for (let i = 0; i < seqs.length; ++i) {
+      assertIsQuantizedSequence(seqs[i]);
+      if (seqs[i].quantizationInfo.stepsPerQuarter !==
+          seqs[0].quantizationInfo.stepsPerQuarter) {
+        throw new Error('Not all sequences have the same quantizationInfo');
+      }
+    }
+    return concatenateHelper(
+        seqs, 'totalQuantizedSteps', 'quantizedStartStep', 'quantizedEndStep',
+        individualDuration);
+  } else {
+    return concatenateHelper(
+        seqs, 'totalTime', 'startTime', 'endTime', individualDuration);
+  }
+}
+
+/**
+ * Trim notes from a NoteSequence to lie within a specified time range.
+ * Notes starting before `start` are not included. Notes ending after
+ * `end` are not included, unless `truncateEndNotes` is true.
+ * @param ns The NoteSequence for which to trim notes.
+ * @param start The time after which all notes should begin. This should be
+ * either seconds (if ns is unquantized), or a quantized step (if ns is
+ * quantized).
+ * @param end The time before which all notes should end. This should be
+ * either seconds (if ns is unquantized), or a quantized step (if ns is
+ * quantized).
+ * @param truncateEndNotes Optional. If true, then notes starting before
+ * the end time but ending after it will be included and truncated.
+ * @returns A new NoteSequence with all notes trimmed to lie between `start`
+ * and `end`, and time-shifted to begin at 0.
+ */
+export function trim(
+    ns: INoteSequence, start: number, end: number, truncateEndNotes?: boolean) {
+  return isQuantizedSequence(ns) ?
+      trimHelper(
+          ns, start, end, 'totalQuantizedSteps', 'quantizedStartStep',
+          'quantizedEndStep', truncateEndNotes) :
+      trimHelper(
+          ns, start, end, 'totalTime', 'startTime', 'endTime',
+          truncateEndNotes);
+}
+
+function concatenateHelper(
+    seqs: INoteSequence[], totalKey: 'totalQuantizedSteps'|'totalTime',
+    startKey: 'startTime'|'quantizedStartStep',
+    endKey: 'endTime'|'quantizedEndStep',
+    individualDuration?: number[]): NoteSequence {
+  let concatSeq: NoteSequence;
```

Comment:
you should probably clone the first one to get the tempos, quantinfo, etc

Reply:
I do, see the (i === 0) branch below.